src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/SyntaxNodeExtensions.cs

```diff
@@ -92,5 +92,129 @@ public static Range GetRange(this SyntaxNode node, RazorSourceDocument source)
+            }
+
+            if (firstToken is null && lastToken is null)
+            {
+                return null;
+            }
+
+            var startPositionSpan = GetLinePositionSpan(firstToken, source, node.SpanStart);
+            var endPositionSpan = GetLinePositionSpan(lastToken, source, node.Span.Start);
+
+            var range = new Range(
+                new Position(startPositionSpan.Start.Line, startPositionSpan.Start.Character),
+                new Position(endPositionSpan.End.Line, endPositionSpan.End.Character));
+
+            return range;
+
+            // This is needed because SyntaxToken positions taken from GetTokens
+            // are relative to their parent node and not to the document.
+            static LinePositionSpan GetLinePositionSpan(SyntaxNode node, RazorSourceDocument source, int parentStart)
+            {
+                if (node is null)
+                {
+                    throw new ArgumentNullException(nameof(node));
+                }
+
+                if (source is null)
+                {
+                    throw new ArgumentNullException(nameof(source));
+                }
+
+                var start = node.Position + parentStart;
+                var end = node.EndPosition + parentStart;
+
+                if (start == source.Length && node.FullWidth == 0)
+                {
+                    // Marker symbol at the end of the document.
+                    var location = node.GetSourceLocation(source);
+                    var position = GetLinePosition(location);
+                    return new LinePositionSpan(position, position);
+                }
+
+                var startLocation = source.Lines.GetLocation(start);
+                var endLocation = source.Lines.GetLocation(end);
+                var startPosition = GetLinePosition(startLocation);
+                var endPosition = GetLinePosition(endLocation);
+
+                return new LinePositionSpan(startPosition, endPosition);
+
+                static LinePosition GetLinePosition(SourceLocation location)
+                {
+                    return new LinePosition(location.LineIndex, location.CharacterIndex);
+                }
+            }
+        }
+
+        public static int GetLeadingWhitespaceLength(this SyntaxNode node)
+        {
+            var tokens = node.GetTokens();
+            for (var i = 0; i < tokens.Count; i++)
+            {
+                var token = tokens[i];
+                if (!IsWhitespace(token))
+                {
+                    return i;
+                }
+            }
+
+            return 0;
+        }
+
+        public static int GetTrailingWhitespaceLength(this SyntaxNode node)
+        {
+            var tokens = node.GetTokens();
+            for (var i = 0;  i < tokens.Count; i++)
+            {
+                var token = tokens[(tokens.Count - 1) - i];
+                if (!IsWhitespace(token))
+                {
+                    return i;
```

Comment:
Tokens are always one character long? Does a newline (`\r\n`) come back as one token or two?

Reply:
You're correct that these calculations aren't quite right. A `\t` needs to be counted as Options.TabSize, NewLine needs to reset the count to 0 etc.